issues-113 add batch support for big-query by using multi values inserts

YaroslavTir opened the pull request:
Fixes #113

schuemie commented on java/org/ohdsi/databaseConnector/BatchedInsert.java:
```diff
@@ -53,53 +46,16 @@ public void executeBatch() {
+
+			List<List<Object>> rows = getListOfRows();
+			System.out.println("----------------sql: ");
+			System.out.println(sql);
```
Why is this written to the console?

schuemie commented on java/org/ohdsi/databaseConnector/BatchedInsert.java:
```diff
@@ -115,6 +71,56 @@ public void executeBatch() {
+		}
+	}
+
+	private List<List<Object>> getListOfRows() {
```
We expect there to be few columns, but many rows. Each column should be of a single type, so for memory-efficiency each column is now stored as a (non-dynamic) array.

You are proposing to copy that to an intermediate structure that is a List of a List of Objects, Each of those comes with memory overhead (the Object class alone needs to store the 64-bit pointer to each value). Not only are you creating an entire unnecessary copy of the (potentially very large) data object, you are doing it in a format that is orders of magnitude less efficient in memory. This is not acceptable.

YaroslavTir replied:
@schuemie , thank you for valuable remark. Now I see why did you use columns approach instead of raws. 
I am not arguing with you here, I just want to find out the best solution. From my point of view, readability and maintainability are more important than small overhead and tbw I would remove arrays completely and use the lists instead. I made a small project out of curiosity where I measured memory footprint. of course, it is not neat code, but it shows that overhead with a collection of 1M element gives the overhead ~10%.
probably in case of using more heavy data, the overhead would be less, anyway I think eventually you are right and I am going to fix it tomorrow

schuemie replied:
Your test code is about a situation where both the array and list contain strings, which are objects. Try again, now comparing int[] versus List<Integer>, and you will see a much larger difference.

YaroslavTir replied:
I get rid of `List<List<>>`. I checked basic cases, run analyses that use batch load, but I don't think this is proper testing.

>our test code is about a situation where both the array and list contain strings

this is right, but eventually, we wrap all primitive values to objects before putting them into prepareStatement, so it would not make any difference. 
I probably play with this test code later, just for self-education, but for now, I left arrays over lists.

schuemie commented:
You're still making unnecessary copies of the data (and making the code unnecessarily complex). You're also batching the regular load in Java, even though it is already batched in R.

Note that we have very little Java heap space to work with, and we have no control over the size of the JVM, as it may already have been created by a previous process. If we're not absolutely stingy with memory, we will get intermittent out-of-heap messages (which happened in the past).

Here's what I propose:

- Create a new function in Java in the existing class that looks almost identical to https://github.com/OHDSI/DatabaseConnector/blob/master/java/org/ohdsi/databaseConnector/BatchedInsert.java#L40 , except it does the values trick, and it does the mini-batching without creating copies of the data. (could be achieved by just adding an outer loop)
- Make R call the alternative Java function for BigQuery here: https://github.com/OHDSI/DatabaseConnector/blob/master/R/InsertTable.R#L435

YaroslavTir commented:
> You're also batching the regular load in Java, even though it is already batched in R
thank you to point this, part. fixed

> You're still making unnecessary copies of the data (and making the code unnecessarily complex).

I don't see where I do copies. `DataLoader` has reference to original columns, this is a reference, not a copy.  I do create inster-values query string in multiValueLoad, but this is necessary.

> Note that we have very little Java heap space to work with, and we have no control over the size of the JVM, as it may already have been created by a previous process. If we're not absolutely stingy with memory, we will get intermittent out-of-heap messages (which happened in the past).

ok, I’ll keep it in my mind that this is one of the main criteria

>Here's what I propose:

This way we ended up with code duplications. that can be an issue in the future. I cleaned up the code. It is straightforward now (from my code of view)

schuemie commented:
I'm sorry, I thought my instructions were clear, but you still made more than the two changes I proposed.
